src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -855,37 +889,122 @@ Node BvInverter::getInversionSkolemFor( Node cond, TypeNode tn ) {
+        unsigned rmod = 0; // TODO : randomize?
+        for( unsigned i=0; i<lit.getNumChildren(); i++ ){
+          unsigned ii = ( i + rmod )%lit.getNumChildren();
+          Node litc = getPathToPv( lit[ii], pv, sv, path, visited );
+          if( !litc.isNull() ){
+            // path is outermost term index last
+            path.push_back( ii );
+            std::vector< Node > children;
+            if( lit.getMetaKind() == kind::metakind::PARAMETERIZED ){
+              children.push_back( lit.getOperator() );
+            }
+            for( unsigned j=0; j<lit.getNumChildren(); j++ ){
+              children.push_back( j==ii ? litc : lit[j] );
+            }
+            return NodeManager::currentNM()->mkNode( lit.getKind(), children );
           }
-          return NodeManager::currentNM()->mkNode( lit.getKind(), children );
         }
       }
     }
   }
   return Node::null();
 }
 
+Node BvInverter::eliminateSkolemFunctions( TNode n, std::vector< Node >& side_conditions, std::unordered_map< TNode, Node, TNodeHashFunction >& visited ) {
+  std::unordered_map< TNode, Node, TNodeHashFunction >::iterator it = visited.find( n );
+  if( it==visited.end() ){
+    Trace("bv-invert-debug") << "eliminateSkolemFunctions from " << n << "..." << std::endl;
+    Node ret = n;
+    bool childChanged = false;
+    std::vector< Node > children;
+    if( n.getMetaKind() == kind::metakind::PARAMETERIZED ){
+      children.push_back( n.getOperator() );
+    }
+    for( unsigned i=0; i<n.getNumChildren(); i++ ){
+      Node nc = eliminateSkolemFunctions( n[i], side_conditions, visited );
+      childChanged = childChanged || n[i]!=nc;
+      children.push_back( nc );
+    } 
+    if( childChanged ){
+      ret = NodeManager::currentNM()->mkNode( n.getKind(), children );
+    }
+    // now, check if it is a skolem function
+    if( ret.getKind()==APPLY_UF ){
+      Node op = ret.getOperator();
+      TypeNode tnp = op.getType();
+      // is this a skolem function?
+      std::map< TypeNode, Node >::iterator its = d_solve_var.find( tnp );
+      if( its!=d_solve_var.end() && its->second==op ){
+        Assert( ret.getNumChildren()==1 );
+        Assert( ret[0].getType().isBoolean() );
+        
+        Node cond = ret[0];
+        // must rewrite now to ensure we lookup the correct skolem 
+        cond = Rewriter::rewrite( cond );
+        
+        // if so, we replace by the (finalized) skolem variable
+        // Notice that since we are post-rewriting, skolem functions are already eliminated from cond
+        ret = getInversionSkolemFor( cond, ret.getType() );
+        
+        // also must add (substituted) side condition to vector
+        // substitute ( solve variable -> inversion skolem )
+        TNode solve_var = getSolveVariable( ret.getType() );
+        TNode tret = ret;
+        cond = cond.substitute( solve_var, tret );
+        if( std::find( side_conditions.begin(), side_conditions.end(), cond )==side_conditions.end() ){
```

Comment:
Using a set for `side_contitions` would be better here, so that we don't have to iterate over the vector for checking if the condition is already added.